# sklearn Classification_report in Python (with Scikit-Learn)

By: jcchouinard.com

-----

## What is a Classification Report

A Classification Report is used to compute the accuracy of a classification model based on the values from the confusion matrix. 

### Sklearn `classification_report()`

In Scikit-learn, `classification_report()` function from `sklearn.metrics` module is used to return a classification report.


In [1]:
# Import necessary libraries
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
import pandas as pd

# Load the dataset
X, y = load_breast_cancer(return_X_y=True)

# split taining and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

# train the model
knn = KNeighborsClassifier(n_neighbors=8)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)

# compute accuracy of the model
knn.score(X_test, y_test)

0.9298245614035088

In [3]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.89      0.90        64
           1       0.94      0.95      0.94       107

    accuracy                           0.93       171
   macro avg       0.93      0.92      0.92       171
weighted avg       0.93      0.93      0.93       171



## Primer on Confusion Matrix

The confusion matrix returns metrics used to make calculations in the classification report.

- `True positives (TP)`: predicted true and it is true. Predicted that someone `is sick` and the person `is sick`.
- `False positives (FP)`: predicted false and it is false. Predicted that someone `is not sick` and the person `is not sick`.
- `True negatives (TN)`: predicted True and it is false. Predicted that someone `is sick` and the person `is not sick`.
- `False negatives (FN)`: predicted false and it is true. Predicted that someone `is not sick` and the person `is sick`.

![](https://www.jcchouinard.com/wp-content/uploads/2023/09/confusion-matrix.jpg)



## Metrics in Classification Report

In [4]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.89      0.90        64
           1       0.94      0.95      0.94       107

    accuracy                           0.93       171
   macro avg       0.93      0.92      0.92       171
weighted avg       0.93      0.93      0.93       171



### Columns of classification_report
- Precision: When we predict True, how often is it true? Higher = Better
- Recall (sensitivity): What percentage of the True values did we predict? Higher = Better
- F1-Score Combines precision and recall. Useful when opposite scores in precision and recall. Higher = Better
- support: number of occurrences of each class in your y_test

### Rows of classification_report
- 0, 1: Classes (Benign, Malignant)
- Accuracy: How often are we predicting the right outcome? Higher = Better
- macro avg: Average of metrics across all classes, treating each class equally. (precision_a + precision_b) / 2
- weighted avg: Class metrics weighted by class size. (precision_a * support_a) + (precision_b * support_b) / total_support


### Recap

| Metric                | What it is                                             | Sklearn’s Metric Method                        |
|-----------------------|--------------------------------------------------------|------------------------------------------------|
| Accuracy              | (true positive + true negative) / total predictions    | `metrics.accuracy_score(true, pred)`           |
| Precision             | true positive / (true positive + false positive)       | `metrics.precision_score(true, pred)`          |
| Recall (sensitivity)  | true positive / (true positive + false negative)       | `metrics.recall_score(true, pred)`             |
| F1-Score              | 2 * (precision * recall) / (precision + recall)        | `metrics.f1_score(true, pred)`                 |
| Specificity           | true negative / (true negative + false positive)       | `metrics.recall_score(true, pred, pos_label=0)`|

## Example with Confusion Matrix

In [6]:
from sklearn.metrics import confusion_matrix
 
cm = confusion_matrix(y_test,y_pred)
cm

array([[ 57,   7],
       [  5, 102]])

In [14]:
tn = cm[0][0] # predicted IS SICK and person IS NOT SICK 
fp = cm[0][1] # predicted IS NOT SICK and person IS NOT SICK
fn = cm[1][0] # predicted IS NOT SICK and person IS SICK 
tp = cm[1][1] # predicted IS SICK and person IS SICK
print(tn, fp, fn, tp)

57 7 5 102


### Compute Metrics from Confusion Matrix

In [17]:
accuracy = (tp + tn) / (tp + fp + fn + tn)
precision = tp / (tp + fp)
recall = tp / (tp + fn)
f1_score = 2 * (precision * recall) / (precision + recall)

print(f'accuracy: {accuracy:.2F}')
print(f'precision: {precision:.2F}')
print(f'recall: {recall:.2F}')
print(f'f1_score: {f1_score:.2F}')

accuracy: 0.93
precision: 0.94
recall: 0.95
f1_score: 0.94


In [16]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.89      0.90        64
           1       0.94      0.95      0.94       107

    accuracy                           0.93       171
   macro avg       0.93      0.92      0.92       171
weighted avg       0.93      0.93      0.93       171



In [24]:
print('Support class 0:', tn + fp)
print('Support class 1:',tp + fn)
print('Support all classes:', tp + fp + fn + tn)

Support class 0: 64
Support class 1: 107
Support all classes: 171


## Interpret the Classification Report

In [5]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.89      0.90        64
           1       0.94      0.95      0.94       107

    accuracy                           0.93       171
   macro avg       0.93      0.92      0.92       171
weighted avg       0.93      0.93      0.93       171



- Accuracy: Highlight that 93% accuracy means the model is right 93% of the time.
- Precision vs. Recall (Focus on Class 1 - Malignant): in the critical case where we want to correctly predict Cancer, we want precision (correct positive predictions) and recall (all actual positive)
    - precision (correct positive predictions). When we predict Cancer, how often is it really Cancer? 94% of the time. High precision: Not many Benign cancers were predicted as Malignant
    - recall (capturing all actual positives) is 95%. What percentage of the real cancers did we manage to predict? 95%. High recall: Predicted most Cancers.
    
This shows how well the model performs in identifying cancer cases, which is vital for real-world applications.

- True Positive (TP)	Correct positive identification.
- True Negative (TN)	Correct negative identification.
- False Positive (FP)	Incorrect positive identification.
- False Negative (FN)	Incorrect negative identification.

Detailed report
- Precision (Class 0 - benign): Out of all the times the model predicted "benign," it was right 92% of the time.
- Recall (Class 0 - benign): Out of all actual "benign" cases, the model correctly identified 89%.
- F1-score (Class 0 - benign): Combines precision and recall, showing that the model is 90% effective at predicting "benign" cases.
- Precision (Class 1 - malignant): Out of all the times the model predicted "malignant," it was right 94% of the time.

Overall:

- Accuracy: The model got the predictions right 93% of the time.
- Macro Average: Looks at how well the model does for both classes equally, without considering how many cases there are for each class.
- Weighted Average: Takes into account that there may be more cases in one class than the other, giving a balanced view of performance across all instances.

## Conclusion

Help me and subscribe to this channel.

Stay tuned for my upcoming Python for SEO course.

### [jcchouinard.com](https://www.jcchouinard.com/)
### [youtube.com/@jcchouinard](https://www.youtube.com/@jcchouinard)
### [twitter.com/ChouinardJC](https://twitter.com/ChouinardJC)
### [linkedin.com/in/jeanchristophechouinard](https://www.linkedin.com/in/jeanchristophechouinard)
